rasterio/__init__.py

```diff
@@ -127,37 +127,53 @@ def open(
       | y | = | d  e  f | | r |
       | 1 |   | 0  0  1 | | 1 |
 
-        a: rate of change of X with respect to increasing column, i.e.  pixel width
-        b: rotation, 0 if the raster is oriented "north up"
-        c: X coordinate of the top left corner of the top left pixel
-        d: rotation, 0 if the raster is oriented "north up"
-        e: rate of change of Y with respect to increasing row, usually
-                a negative number (i.e. -1 * pixel height) if north-up.
-        f: Y coordinate of the top left corner of the top left pixel
-
-    A 6-element sequence of the affine transformation
-    matrix coefficients in ``c, a, b, f, d, e`` order,
-    (i.e. GDAL geotransform order) will be accepted until 1.0 (deprecated).
-
-    A virtual filesystem can be specified. The ``vfs`` parameter may be
-    an Apache Commons VFS style string beginning with "zip://" or
+      a: rate of change of X with respect to increasing column,
+         i.e. pixel width
+      b: rotation, 0 if the raster is oriented "north up"
+      c: X coordinate of the top left corner of the top left pixel
+      d: rotation, 0 if the raster is oriented "north up"
+      e: rate of change of Y with respect to increasing row,
+         usually a negative number (i.e. -1 * pixel height) if
+         north-up.
+      f: Y coordinate of the top left corner of the top left pixel
+
+    A 6-element sequence of the affine transformation matrix
+    coefficients in ``c, a, b, f, d, e`` order, (i.e. GDAL
+    geotransform order) will be accepted until 1.0 (deprecated).
+
+    A virtual filesystem can be specified. The ``vfs`` parameter may
+    be an Apache Commons VFS style string beginning with "zip://" or
     "tar://"". In this case, the ``path`` must be an absolute path
     within that container.
 
     """
     if not isinstance(path, string_types):
-        raise TypeError("invalid path: %r" % path)
+        raise TypeError("invalid path: {0!r}".format(path))
     if mode and not isinstance(mode, string_types):
-        raise TypeError("invalid mode: %r" % mode)
+        raise TypeError("invalid mode: {0!r}".format(mode))
     if driver and not isinstance(driver, string_types):
-        raise TypeError("invalid driver: %r" % driver)
-
+        raise TypeError("invalid driver: {0!r}".format(driver))
+    if dtype and not check_dtype(dtype):
+        raise TypeError("invalid dtype: {0!r}".format(dtype))
     if transform:
         transform = guard_transform(transform)
     elif 'affine' in kwargs:
         affine = kwargs.pop('affine')
         transform = guard_transform(affine)
 
+    # If there is no currently active GDAL/AWS environment, create one.
+    defenv()
+
+    # Get AWS credentials if we're attempting to access a raster
+    # on S3.
+    pth, archive, scheme = parse_path(path)
+    if scheme == 's3':
+        Env().get_aws_credentials()
```

Comment:
@sgillies Having some problems with GDAL 2 so I haven't fully confirmed, but I think this will ignore any credentials supplied by an outer `Env()` like:

```python
with rasterio.env.Env(aws_access_key_id='something'):
    with rasterio.open('data.tif') as src:
        pass
```

Instead of setting `AWS_ACCESS_KEY_ID=something` we get whatever value `boto3.Session().get_credentials()` provides.  I think something like `Env(**getenv().options)` would work, but the AWS credentials are not pushed into the GDAL environment until `Env.get_aws_credentials()` is called, which means the outer credentials are almost certainly not present at this point.